compiler/grainformat/reformat.re

```diff
@@ -1045,78 +1048,136 @@ and print_application =
+            ~infixFunction=None,
             first,
           ),
           Doc.rparen,
         ])
-      | PExpApp(fn, _) =>
-        let leftfn = get_function_name(fn);
-        if (infixop(leftfn)) {
-          Doc.concat([
-            Doc.lparen,
-            print_expression(
-              ~parentIsArrow=false,
-              ~endChar=None,
-              ~original_source,
-              ~parent_loc,
-              first,
-            ),
-            Doc.rparen,
-          ]);
-        } else {
-          Doc.concat([
-            Doc.lparen,
-            print_expression(
-              ~parentIsArrow=false,
-              ~endChar=None,
-              ~original_source,
-              ~parent_loc,
-              first,
-            ),
-            Doc.rparen,
-          ]);
-        };
+      | PExpApp(_) =>
+        print_expression(
+          ~parentIsArrow=false,
+          ~endChar=None,
+          ~original_source,
+          ~parent_loc,
+          ~infixFunction=Some(function_name),
+          first,
+        )
+
       | _ =>
         print_expression(
           ~parentIsArrow=false,
           ~endChar=None,
           ~original_source,
           ~parent_loc,
+          ~infixFunction=None,
           first,
         )
       };
 
     let second_brackets =
       switch (second.pexp_desc) {
-      | PExpIf(_)
-      | PExpApp(_) =>
+      | PExpIf(_) =>
         Doc.concat([
           Doc.lparen,
           print_expression(
             ~parentIsArrow=false,
             ~endChar=None,
             ~original_source,
             ~parent_loc,
+            ~infixFunction=None,
             second,
           ),
           Doc.rparen,
         ])
+      | PExpApp(_) =>
+        print_expression(
+          ~parentIsArrow=false,
+          ~endChar=None,
+          ~original_source,
+          ~parent_loc,
+          ~infixFunction=Some(function_name),
+          second,
+        )
       | _ =>
         print_expression(
           ~parentIsArrow=false,
           ~endChar=None,
           ~original_source,
           ~parent_loc,
+          ~infixFunction=None,
           second,
         )
       };
-    Doc.concat([
-      first_brackets,
-      Doc.space,
-      Doc.text(function_name),
-      Doc.space,
-      second_brackets,
-    ]);
+
+    let op_precedence = fn =>
+      switch (fn) {
+      | "*"
+      | "/"
+      | "%" => 120
+      | "+"
+      | "-"
+      | "++" => 110
+      | "<<"
+      | ">>"
+      | ">>>" => 100
+      | "<"
+      | "<="
+      | ">"
+      | ">=" => 90
+      | "=="
+      | "!="
+      | "is"
+      | "isnt" => 80
+      | "&" => 70
+      | "^" => 60
+      | "|" => 50
+      | "&&" => 40
+      | "||" => 30
+      | "_" => 10
+      | _ => 0
+      };
+
+    let override_precedence = (fn2, fn1) => {
+      let left_prec = op_precedence(fn1);
+      let right_prec = op_precedence(fn2);
+
+      if (left_prec < right_prec) {
+        true;
+      } else {
+        false;
+      };
```

Comment:
```suggestion
      left_prec < right_prec
```

Reply:
Oh dear god.   Yes, yours is much better.

Reply:
(That was left over from some debugging)